### PR TITLE
Rewrite open api case classes to normal classes for backwards compatibility

### PR DESCRIPTION
**Pull request by bmeesters**
Partially solves #531

### CODE REVIEW

**Comment by bmeesters** [comment]
Hello @julienrf I have started on id:531 as discussed. I think this is all that has to happen right (but then for every class)? I suppose all case classes in the `OpenApi.scala` file should be done? Just checking, before I made the same mistake over and over for every domain object.

**Comment by julienrf** [comment]
Thanks @bmeesters for starting the discussion :)

>  I think this is all that has to happen

We should also override `hashCode` and `equals`.

> I suppose all case classes in the OpenApi.scala file should be done?

Ideally, yes… We should be able to support any minor evolution of the OpenAPI schema.

There is also the class `EndpointDocs`. If that’s too much work we can split it, let me know what you want to do.

**Comment by bmeesters** [comment]
No problem :slightly_smiling_face: 

How should the `hashCode` look like? `equals` I can imagine. I will probably start with all case classes in the OpenApi file. If it is just these two extra classes I will also do those. Though any more I would like to separate them. Its not complex work, but it is a bit tedious.

**Comment by julienrf** [comment]
We can just follow the pattern used by sbt-contraband, I believe: https://www.scala-sbt.org/contraband/code-generation.html#Record+types

**Comment by bmeesters** [comment]
> We can just follow the pattern used by sbt-contraband, I believe: https://www.scala-sbt.org/contraband/code-generation.html#Record+types

Including the numbers (37, 37, 17) or just adding the hashcodes of the individual fields? I don't really understand how this came to be, which is why I am asking.

**Comment by julienrf** [comment]
Yes, the factors are important (and the parenthesis as well). You can find more information here: https://stackoverflow.com/questions/7370925/what-is-the-standard-idiom-for-implementing-equals-and-hashcode-in-scala

In case you use IntelliJ, it can generate those methods for you.

**Comment by harpocrates** [comment]
Why not just use `MurmurHash3` directly? I think this is anyways what case classes do nowadays, right?

**Comment by julienrf** [comment]
[Indeed](https://github.com/scala/scala/blob/51233da2bc6fb9e4315c8dc9620f213a5e13c888/src/library/scala/runtime/ScalaRunTime.scala#L161), I didn’t know. Yes, we should use that too!

File: openapi/openapi/src/main/scala/endpoints/openapi/model/OpenApi.scala [hunk 1]
```diff
@@ -269,65 +304,467 @@ object OpenApi {
...
+}
+
+object Info {
+
+  def apply(title: String, version: String): Info =
+    new Info(title, version)
+
+}
+
+class PathItem(
```
Comment (julienrf) [inline openapi/openapi/src/main/scala/endpoints/openapi/model/OpenApi.scala:345]: Should we make the constructor `private`?

Comment (bmeesters) [reply openapi/openapi/src/main/scala/endpoints/openapi/model/OpenApi.scala:345]: Yes, I think that makes sense.

**Comment by bmeesters** [comment]
Hey @julienrf, it took a while, but I think this is ready for review now. I didn't do any of the changes regarding `Tag` since that would require logic changes and this is only rewrites of existing classes. I plan to start on that once this is merged. Can you have a look?

**Comment by julienrf** [comment]
Thanks a lot @bmeesters, I’ll review it today.

File: json-schema/json-schema/src/test/scala/endpoints/HashingTest.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import org.scalacheck.Properties
+import org.scalacheck.Prop.forAll
+
+object HashingTest extends Properties("Hashing") {
+
+  property("hash") = forAll { (x: Int, y: Int, s: String, t: String) =>
+    val oneValue =
+      (x == y) == (Hashing.hash(x) == Hashing.hash(y))
+    val twoValues =
+      ((x, s) == (y, t)) == (Hashing.hash(x, s) == Hashing.hash(y, t))
```
Comment (julienrf) [inline json-schema/json-schema/src/test/scala/endpoints/HashingTest.scala:12]: It seems that this syntax doesn’t work with Scala 2.

```suggestion
      ((x, s) == ((y, t))) == (Hashing.hash(x, s) == Hashing.hash(y, t))
```

**Comment by bmeesters** [comment]
> I just have one open question. For types EndpointDocs and CallbackDocs, I’ve defined a apply methods that only take required parameters, and optional parameters (or those for which we have a sensible default) can be set by using withXxx methods. Maybe it would be more regular to follow this scheme everywhere… what do you think? The drawback is that this might create incompatibilities with the current API (see the @deprecated methods I’ve added).

I think that makes sense. The breakage is so minor that I don't expect anyone to have major troubles moving to the new scheme. I'll do one more commit to add that as well.

**Comment by julienrf** [comment]
There is one last thing: we should make the classes `final` as well.

**Comment by bmeesters** [comment]
Right, forgot about that. Done now.

**Pull request merged**